src/main/kotlin/frc/team449/subsystems/drive/swerve/SwerveDrive.kt

```diff
@@ -45,6 +51,42 @@ open class SwerveDrive(
 
   var desiredSpeeds: ChassisSpeeds = ChassisSpeeds()
 
+  var desiredAngle = 0.0
+  var desiredOmega = 0.0
+
+  // Removed magic numbers
+  private val xController: PIDController
+    get() = PIDController(AutoConstants.DEFAULT_X_KP, 0.0, 0.0)
+  private val yController: PIDController
+    get() = PIDController(AutoConstants.DEFAULT_Y_KP, 0.0, 0.0)
+  private val headingController: PIDController
+    get() = PIDController(AutoConstants.DEFAULT_ROTATION_KP, 0.0, 0.0)
+
+  init {
+    headingController.enableContinuousInput(-Math.PI, Math.PI)
+  }
+  fun followTrajectory(
+    robot: Robot,
+    sample: SwerveSample
+  ) {
+    desiredAngle = MathUtil.angleModulus(sample.heading)
+    desiredOmega = sample.omega
+    val speeds = ChassisSpeeds(
+      sample.vx + xController.calculate(robot.poseSubsystem.pose.x, sample.x),
+      sample.vy + yController.calculate(robot.poseSubsystem.pose.y, sample.y),
+      sample.omega + headingController.calculate(
+        robot.poseSubsystem.pose.rotation.minus(Rotation2d.fromRadians(MathUtil.angleModulus(sample.heading))).radians
+      )
+    )
+    val newSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(
+      speeds,
+      robot.poseSubsystem.heading
+    )
+
+    // Apply the generated speeds
+    robot.drive.set(newSpeeds)
```

Comment:
Why are we going through the robot class to access this call? We're in the drive class, just call set.